File: 934.最短的桥.py
#
# @lc app=leetcode.cn id=934 lang=python
#
# [934] 最短的桥
#

# @lc code=start
class Solution(object):
    def shortestBridge(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        N = len(grid)

        def DFS(x, y):
            if x<0 or y<0 or x>N-1 or y>N-1: return
            if grid[x][y] != 1: return
            if grid[x][y] == 1:
                grid[x][y] = 2
                S.add((x, y))
            DFS(x+1, y)
            DFS(x-1, y)
            DFS(x, y+1)
            DFS(x, y-1)

        def BFS():
            queue = []
            visited = set()
            for s in S:
                queue.append((s,0))
                visited.add(s)
            while queue:
                (i, j), step = queue.pop(0)
                if grid[i][j] == 1: return step-1
                for x, y in [(i-1,j),(i,j-1),(i+1,j),(i,j+1)]:
                    if x>=0 and y>=0 and x<=N-1 and y<=N-1:
                        if (x,y) not in visited:
                            visited.add((x,y))
                            queue.append(((x,y),step+1))
        
        S = set()
        flag = False
        for i in range(N):
            for j in range(N):
                if grid[i][j] == 1:
                    DFS(i, j)
                    flag = True
                    break
            if flag: break
        
        return BFS()
# @lc code=end

File: javascript/934.最短的桥.js
/*
 * @lc app=leetcode.cn id=934 lang=javascript
 *
 * [934] 最短的桥
 * 
 * 注意啥叫dfs啥叫bfs：
 *  bfs是找到一个解就可以了，第一个岛向外延伸只要碰到第二个岛就停，这就叫bfs
 *  dfs是找到所有解，把第一个岛所有都标为2这就是dfs
 * 
 * 注意js用set查找会超时，只能变成二维bit数组
 * 把所有可能起始点都直接放在queue中做会更快，每个点单独做完再取最小值浪费了很多计算
 */

// @lc code=start
/**
 * @param {number[][]} grid
 * @return {number}
 */
var shortestBridge = function(grid) {
    var n = grid.length;
    var nodes = [];

    var dfs = function(i, j) {
        if (i<0 || i>=n || j<0 || j>=n) { return; }
        if (grid[i][j]==0 || grid[i][j]==2) { return; }

        grid[i][j] = 2;
        nodes.push([i, j]);
        dfs(i+1, j);
        dfs(i-1, j);
        dfs(i, j+1);
        dfs(i, j-1);
    };

    var bfs = function(i, j) {
        var queue = [];
        var visited = Array.from(new Array(n), ()=>new Array(n).fill(false));
        for (var node of nodes) {
            queue.push([node[0], node[1], 0]);
            visited[node[0]][node[1]] = true;
        }
        
        while (queue.length) {
            var data = queue.shift();
            var x=data[0], y=data[1], cost=data[2];
            if (grid[x][y] == 1) {
                return cost-1;
            }
            for (var dir of [[x+1,y], [x-1,y], [x,y-1], [x,y+1]]) {
                var dx=dir[0], dy=dir[1];
                if (dx>-1 && dx<n && dy>-1 && dy<n && !visited[dx][dy]) {
                    queue.push([dx, dy, cost+1]);
                    visited[dx][dy] = true;
                }
            }
        }
    };


    for (var i=0; i<n; i++) {
        for (var j=0; j<n; j++) {
            if (grid[i][j] == 1) {
                dfs(i, j);

                return bfs();
            }
        }
    }
};
// @lc code=end

